### PR TITLE
Convert some Java classes to Kotlin (and add basic unit tests for resource utils)

### PR DESCRIPTION
**Pull request by jingtang10**

### CODE REVIEW

File: fhirengine/src/main/java/com/google/fhirengine/ResourceAlreadyExistsException.kt [hunk 1]
```diff
@@ -14,24 +14,11 @@
...
-  public String getType() {
-    return type;
-  }
-
-  public String getId() {
-    return id;
-  }
-}
+class ResourceAlreadyExistsException(
+  type: String,
```
Comment (yigit) [inline fhirengine/src/main/java/com/google/fhirengine/ResourceAlreadyExistsException.kt:21]: these need to be `val`'s. (except for cause)

File: fhirengine/src/main/java/com/google/fhirengine/ResourceNotFoundException.kt [hunk 2]
```diff
@@ -14,24 +14,11 @@
...
-  public String getType() {
-    return type;
-  }
-
-  public String getId() {
-    return id;
-  }
-}
+class ResourceNotFoundException(
+  type: String,
```
Comment (yigit) [inline fhirengine/src/main/java/com/google/fhirengine/ResourceNotFoundException.kt:21]: ditto

File: fhirengine/src/main/java/com/google/fhirengine/resource/Resources.kt [hunk 3]
```diff
@@ -0,0 +1,50 @@
...
+ */
+
+package com.google.fhirengine.resource
+
+import java.lang.reflect.InvocationTargetException
+import org.hl7.fhir.r4.model.Resource
+import org.hl7.fhir.r4.model.ResourceType
+
+/** The HAPI Fhir package prefix for R4 resources. */
+val R4_RESOURCE_PACKAGE_PREFIX = "org.hl7.fhir.r4.model."
```
Comment (yigit) [inline fhirengine/src/main/java/com/google/fhirengine/resource/Resources.kt:24]: i see the previous java class was public but i think these are all internal functionality right?
if so, we should mark these `internal` (everything in this file)

File: fhirengine/src/main/java/com/google/fhirengine/resource/Resources.kt [hunk 4]
```diff
@@ -0,0 +1,50 @@
...
+        throw IllegalArgumentException("Cannot resolve resource type for " + clazz.getName(), e)
+    } catch (e: InstantiationException) {
+        throw IllegalArgumentException("Cannot resolve resource type for " + clazz.getName(), e)
+    } catch (e: InvocationTargetException) {
+        throw IllegalArgumentException("Cannot resolve resource type for " + clazz.getName(), e)
+    }
+}
+
+/** Returns the {@link Class} object for the resource type. */
+fun <R : Resource> getResourceClass(resourceType: String): Class<R> {
```
Comment (yigit) [inline fhirengine/src/main/java/com/google/fhirengine/resource/Resources.kt:42]: we may want to move this to KClass instead but does not need to be in this PR

**Review by yigit — CHANGES_REQUESTED** [review]

File: fhirengine/src/main/java/com/google/fhirengine/ResourceAlreadyExistsException.kt (hunk 1 above)
Comment (jingtang10) [reply fhirengine/src/main/java/com/google/fhirengine/ResourceAlreadyExistsException.kt:21]: why? we don't need them after the constructor right?

File: fhirengine/src/main/java/com/google/fhirengine/ResourceNotFoundException.kt (hunk 2 above)
Comment (jingtang10) [reply fhirengine/src/main/java/com/google/fhirengine/ResourceNotFoundException.kt:21]: ditto

File: fhirengine/src/main/java/com/google/fhirengine/resource/Resources.kt (hunk 3 above)
Comment (jingtang10) [reply fhirengine/src/main/java/com/google/fhirengine/resource/Resources.kt]: yeah thanks for the catch!

File: fhirengine/src/main/java/com/google/fhirengine/resource/Resources.kt (hunk 4 above)
Comment (jingtang10) [reply fhirengine/src/main/java/com/google/fhirengine/resource/Resources.kt]: ack. i'll do that in a following pr.

**Comment by jingtang10** [comment]
rebased. PTAL! thx! @yigit

File: fhirengine/src/androidTest/java/com/google/fhirengine/db/impl/DatabaseImplTest.kt [hunk 5]
```diff
@@ -158,4 +158,22 @@ class DatabaseImplTest {
             TEST_PATIENT_2.setGender(Enumerations.AdministrativeGender.MALE)
         }
     }
+
+    private companion object {
```
Comment (deepankarb) [inline fhirengine/src/androidTest/java/com/google/fhirengine/db/impl/DatabaseImplTest.kt:162]: Duplicate code? There's another `companion` right above.

File: fhirengine/src/androidTest/java/com/google/fhirengine/db/impl/DatabaseImplTest.java [hunk 6]
```diff
@@ -0,0 +1,137 @@
+/*
```
Comment (deepankarb) [inline fhirengine/src/androidTest/java/com/google/fhirengine/db/impl/DatabaseImplTest.java:1]: This should be removed.

**Review by deepankarb — CHANGES_REQUESTED** [review]
Empty file named 'Task' can be removed. Left a couple of comments. Thanks!

Also please take a look at spotless failure in CI: https://github.com/google/android-fhir/pull/113/checks?check_run_id=1487953151#step:10:27

File: fhirengine/src/main/java/com/google/fhirengine/ResourceAlreadyExistsException.kt (hunk 1 above)
Comment (yigit) [reply fhirengine/src/main/java/com/google/fhirengine/ResourceAlreadyExistsException.kt:21]: there are public getters/setters for it in the java version. i thought you wanted to keep them as properties ? with his, they cannot be read back anymore.

Comment (jingtang10) [reply fhirengine/src/main/java/com/google/fhirengine/ResourceAlreadyExistsException.kt]: yeah you're right. if we don't expose these the class is a bit pointless.

**Review by yigit — APPROVED** [review]

**Pull request merged**